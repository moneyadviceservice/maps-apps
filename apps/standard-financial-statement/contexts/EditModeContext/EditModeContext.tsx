import {
  createContext,
  FormEvent,
  ReactNode,
  useContext,
  useMemo,
  useState,
} from 'react';

import { useRouter } from 'next/router';

import { organisationFormObject } from '../../utils/organisations/formatFormObjects/organisationFormObject';

interface EditModeContextProps {
  isEditMode: boolean;
  setIsEditMode: (val: boolean) => void;
  handleCancel: () => void;
  handleSave: (e: FormEvent<HTMLFormElement>) => void;
}

const EditModeContext = createContext<EditModeContextProps | null>(null);

export const useEditMode = () => {
  const context = useContext(EditModeContext);
  if (!context)
    throw new Error('useEditMode must be used within ToggleEditMode');
  return context;
};

export const ToggleEditProvider = ({
  licence_number,
  children,
}: {
  licence_number?: string;
  children: ReactNode;
}) => {
  const router = useRouter();

  const [isEditMode, setIsEditMode] = useState(false);

  const handleCancel = () => setIsEditMode(false);
  const handleSave = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    try {
      if (!licence_number) {
        throw new Error('Missing required fields to update the record');
      }

      const form = e?.target as HTMLFormElement;
      const formData = new FormData(form);

      const payload = organisationFormObject(formData);

      const res = await fetch('/api/update-organisation', {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ licence_number, payload }),
      });

      const result = await res.json();
      if (!res.ok) {
        throw new Error(result.error ?? 'Failed to update organisation');
      }

      setIsEditMode(false);

      router.replace(router.asPath);
    } catch (err) {
      console.error('Submission error:', err);
    }
  };

  const contextValue = useMemo(
    () => ({
      isEditMode,
      setIsEditMode,
      handleCancel,
      handleSave,
    }),
    [isEditMode],
  );

  return (
    <EditModeContext.Provider value={contextValue}>
      {isEditMode ? <form onSubmit={handleSave}>{children}</form> : children}
    </EditModeContext.Provider>
  );
};
