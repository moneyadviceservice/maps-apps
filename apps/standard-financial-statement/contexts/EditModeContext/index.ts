export { ToggleEditProvider, useEditMode } from './EditModeContext';
