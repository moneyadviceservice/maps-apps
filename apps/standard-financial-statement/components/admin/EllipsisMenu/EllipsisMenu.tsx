import { useEffect, useRef, useState } from 'react';

import { useRouter } from 'next/navigation';

import { Modal } from 'components/Modal';
import { Organisation } from 'types/Organisations';

import { Button } from '@maps-react/common/components/Button';

import { useEditMode } from '../../../contexts/EditModeContext';
import { ActionModalContent } from '../profile/ManageLicence/ActionModalContent';

type Props = {
  data: Organisation;
};

export const EllipsisMenu = ({ data }: Props) => {
  const router = useRouter();
  const { setIsEditMode } = useEditMode();

  const menuRef = useRef<HTMLDivElement>(null);

  const [showMenu, setShowMenu] = useState(false);
  const [modalOpen, setModalOpen] = useState(false);
  const [error, setError] = useState<string>();

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (menuRef.current && !menuRef.current.contains(event.target as Node)) {
        setShowMenu(false);
      }
    };

    if (showMenu) {
      document.addEventListener('mousedown', handleClickOutside);
    }

    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [showMenu]);

  const deleteOrganisation = async () => {
    const cosmosDeleteParams = {
      licence_number: data.licence_number,
    };

    try {
      await fetch(`/api/delete-organisation`, {
        method: 'DELETE',
        body: JSON.stringify(cosmosDeleteParams),
      });
    } catch (err) {
      console.error((err as Error).message);
      setError((err as Error).message);

      return err;
    }

    setModalOpen(false);
    router.push('/admin/dashboard');
  };

  return (
    <div ref={menuRef} className="relative">
      <Button
        onClick={() => setShowMenu(!showMenu)}
        target="_blank"
        variant="secondary"
        className="flex justify-center h-10 w-10 text-xl font-bold hover:bg-gray-100 rounded text-gray-400"
        data-testid="ellipsis-menu"
      >
        <span className="-mt-2">...</span>
      </Button>
      {showMenu && (
        <div className="absolute right-0 mt-2 bg-white border border-gray-400 rounded shadow-lg w-52">
          <button
            className="w-full text-left px-4 py-2 hover:bg-gray-200 rounded-t"
            onClick={() => {
              setIsEditMode(true);
              setShowMenu(false);
            }}
          >
            Edit organisation
          </button>
          <button
            className="w-full text-left px-4 py-2 bg-gray-100 hover:bg-gray-200 rounded-b"
            onClick={() => {
              setModalOpen(true);
              setShowMenu(false);
            }}
            data-testid="ellipsis-delete-button"
          >
            Delete organisation
          </button>
        </div>
      )}
      {modalOpen && (
        <Modal
          isOpen={modalOpen}
          onClose={() => setModalOpen(false)}
          data-testid="modal-container"
        >
          <ActionModalContent
            action={'delete'}
            orgName={data.name}
            onCancel={() => setModalOpen(false)}
            onConfirm={deleteOrganisation}
            error={error}
          />
        </Modal>
      )}
    </div>
  );
};
