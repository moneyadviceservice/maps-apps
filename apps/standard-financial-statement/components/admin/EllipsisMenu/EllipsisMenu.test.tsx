import { ReactNode } from 'react';

import { useRouter } from 'next/navigation';

import { fireEvent, render, screen, waitFor } from '@testing-library/react';

import { useEditMode } from '../../../contexts/EditModeContext';
import { EllipsisMenu } from './EllipsisMenu';

import '@testing-library/jest-dom';

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}));

jest.mock('../../../contexts/EditModeContext', () => ({
  useEditMode: jest.fn(),
}));

jest.mock('../../Modal', () => ({
  Modal: ({ children, isOpen }: { isOpen: boolean; children: ReactNode }) =>
    isOpen ? <div data-testid="modal-dialog">{children}</div> : null,
}));

const mockPush = jest.fn();
const mockSetIsEditMode = jest.fn();

const organisationMock = {
  id: '1',
  name: 'Test Org',
  licence_number: 12345,
  type: { title: '' },
  licence_status: '',
  sfs_live: true,
  created: '',
  modified: '',
  users: [{ email: 'user1@test.com' }, { email: 'user2@test.com' }],
};

describe('EllipsisMenu component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    (useRouter as jest.Mock).mockReturnValue({ push: mockPush });
    (useEditMode as jest.Mock).mockReturnValue({
      setIsEditMode: mockSetIsEditMode,
    });
    global.fetch = jest.fn();
  });

  const openDeleteModal = () => {
    render(<EllipsisMenu data={organisationMock} />);
    fireEvent.click(screen.getByTestId('ellipsis-menu'));
    fireEvent.click(screen.getByTestId('ellipsis-delete-button'));
  };

  it('should render button initially', () => {
    render(<EllipsisMenu data={organisationMock} />);

    const button = screen.getByRole('button', { name: '...' });
    expect(button).toBeInTheDocument();
  });

  it('should show menu when button is clicked', () => {
    render(<EllipsisMenu data={organisationMock} />);

    const button = screen.getByRole('button', { name: '...' });
    fireEvent.click(button);

    const menu = screen.getByRole('button', { name: /edit organisation/i });
    expect(menu).toBeInTheDocument();
  });

  it('should hide menu when clicking outside', () => {
    render(<EllipsisMenu data={organisationMock} />);

    const button = screen.getByRole('button', { name: '...' });
    fireEvent.click(button);

    const menu = screen.getByRole('button', { name: /edit organisation/i });
    expect(menu).toBeInTheDocument();

    fireEvent.mouseDown(document.body);

    expect(menu).not.toBeInTheDocument();
  });

  it('should call setIsEditMode when "Edit organisation" is clicked', () => {
    render(<EllipsisMenu data={organisationMock} />);

    const button = screen.getByRole('button', { name: '...' });
    fireEvent.click(button);

    const editButton = screen.getByRole('button', {
      name: /edit organisation/i,
    });
    fireEvent.click(editButton);

    expect(mockSetIsEditMode).toHaveBeenCalledWith(true);
  });

  it('opens the modal on delete click', () => {
    openDeleteModal();
    expect(screen.getByTestId('modal-dialog')).toBeInTheDocument();
    expect(screen.getByTestId('modal-primary-button')).toBeInTheDocument();
  });

  it('calls both delete APIs and redirects on confirm after successful deletion', async () => {
    (fetch as jest.Mock)
      .mockResolvedValueOnce({ ok: true })
      .mockResolvedValueOnce({ ok: true });

    openDeleteModal();

    fireEvent.click(screen.getByTestId('modal-primary-button'));

    await waitFor(() => {
      expect(fetch).toHaveBeenCalledWith('/api/delete-organisation', {
        method: 'DELETE',
        body: JSON.stringify({
          licence_number: organisationMock.licence_number,
        }),
      });

      expect(fetch).toHaveBeenCalledWith('/api/users/delete', {
        method: 'DELETE',
        body: JSON.stringify({ users: organisationMock.users }),
      });

      expect(mockPush).toHaveBeenCalledWith('/admin/dashboard');
    });
  });

  it('displays error if cosmos delete API fails', async () => {
    (fetch as jest.Mock)
      .mockRejectedValueOnce(new Error('Cosmos delete failed'))
      .mockResolvedValueOnce({ ok: true });

    openDeleteModal();

    fireEvent.click(screen.getByTestId('modal-primary-button'));

    await waitFor(() => {
      expect(fetch).toHaveBeenCalledTimes(1);
      expect(screen.getByText(/cosmos delete failed/i)).toBeInTheDocument();
      expect(mockPush).not.toHaveBeenCalled();
    });
  });

  it('displays error if users delete API fails', async () => {
    (fetch as jest.Mock)
      .mockResolvedValueOnce({ ok: true })
      .mockRejectedValueOnce(new Error('User delete failed'));

    openDeleteModal();

    fireEvent.click(screen.getByTestId('modal-primary-button'));

    await waitFor(() => {
      expect(fetch).toHaveBeenCalledTimes(2);
      expect(screen.getByText(/user delete failed/i)).toBeInTheDocument();
      expect(mockPush).not.toHaveBeenCalled();
    });
  });

  it('matches snapshot', async () => {
    (fetch as jest.Mock)
      .mockResolvedValueOnce({ ok: true })
      .mockResolvedValueOnce({ ok: true });

    const { container } = render(<EllipsisMenu data={organisationMock} />);

    fireEvent.click(screen.getByTestId('ellipsis-menu'));
    fireEvent.click(screen.getByTestId('ellipsis-delete-button'));

    expect(container).toMatchSnapshot();
  });
});
