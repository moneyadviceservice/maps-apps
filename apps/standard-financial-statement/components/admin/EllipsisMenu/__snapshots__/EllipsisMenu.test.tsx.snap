// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EllipsisMenu component matches snapshot 1`] = `
<div>
  <div
    class="relative"
  >
    <button
      class="border border-magenta-500 bg-white hover:border-pink-800 hover:text-pink-800 focus:border-transparent active:text-slate-900 active:outline-blue-700 active:bg-gray-50 active:px-[13px] active:py-[9px] active:border-none t-button relative px-3 py-2 outline-none shadow-bottom-gray items-center gap-2 cursor-pointer disabled:cursor-not-allowed disabled:border-slate-400 disabled:border disabled:bg-transparent disabled:text-gray-400 disabled:shadow-none focus:bg-yellow-400 focus:outline-blue-700 focus:outline-[3px] focus:outline-offset-0 focus:text-gray-800 focus:shadow-none focus:before:content-[""] focus:before:w-[calc(100%+6px)] focus:before:rounded-[0.45rem] focus:before:absolute focus:before:inset-0 focus:before:inset-x-[-3px] focus:before:inset-y-[-3px] focus:before:shadow-bottom-gray active:outline-[3px] active:outline-offset-0 active:shadow-inset active:before:content-[""] active:before:w-[calc(100%+6px)] active:before:rounded-[0.45rem] active:before:absolute active:before:inset-0 active:before:inset-x-[-3px] active:before:inset-y-[-3px] active:before:shadow-bottom-white t-secondary-button flex justify-center h-10 w-10 text-xl font-bold hover:bg-gray-100 rounded text-gray-400"
      data-testid="ellipsis-menu"
      target="_blank"
      type="submit"
    >
      <span />
      <span
        class="-mt-2"
      >
        ...
      </span>
      <span />
    </button>
    <div
      data-testid="modal-dialog"
    >
      <h2
        class="text-3xl md:text-5xl font-bold mb-6 text-blue-700"
        data-testid="modal-heading"
      >
        Are you sure you want to delete this organisation?
      </h2>
      <p
        class="mb-4 text-xl"
        data-testid="modal-content"
      >
         Please confirm whether you would like to delete this organisation. You cannot undo this action.
      </p>
      <div
        class="flex gap-4"
      >
        <button
          class="border-0 text-white active:bg-pink-900 active:outline-yellow-400 active:text-white t-button relative px-3 py-2 font-semibold rounded outline-none shadow-bottom-gray inline-flex items-center justify-between gap-2 cursor-pointer disabled:cursor-not-allowed disabled:border-slate-400 disabled:border disabled:bg-transparent disabled:text-gray-400 disabled:shadow-none focus:bg-yellow-400 focus:outline-blue-700 focus:outline-[3px] focus:outline-offset-0 focus:text-gray-800 focus:shadow-none focus:before:content-[""] focus:before:w-[calc(100%+6px)] focus:before:rounded-[0.45rem] focus:before:absolute focus:before:inset-0 focus:before:inset-x-[-3px] focus:before:inset-y-[-3px] focus:before:shadow-bottom-gray active:outline-[3px] active:outline-offset-0 active:shadow-inset active:before:content-[""] active:before:w-[calc(100%+6px)] active:before:rounded-[0.45rem] active:before:absolute active:before:inset-0 active:before:inset-x-[-3px] active:before:inset-y-[-3px] active:before:shadow-bottom-white t-primary-button bg-red-600 hover:bg-red-700"
          data-testid="modal-primary-button"
          type="submit"
        >
          <span />
          Yes, DELETE this organisation
          <span />
        </button>
        <button
          class="border border-magenta-500 text-magenta-500 bg-white hover:border-pink-800 hover:text-pink-800 hover:bg-gray-50 focus:border-transparent active:text-slate-900 active:outline-blue-700 active:bg-gray-50 active:px-[13px] active:py-[9px] active:border-none t-button relative px-3 py-2 font-semibold rounded outline-none shadow-bottom-gray inline-flex items-center justify-between gap-2 cursor-pointer disabled:cursor-not-allowed disabled:border-slate-400 disabled:border disabled:bg-transparent disabled:text-gray-400 disabled:shadow-none focus:bg-yellow-400 focus:outline-blue-700 focus:outline-[3px] focus:outline-offset-0 focus:text-gray-800 focus:shadow-none focus:before:content-[""] focus:before:w-[calc(100%+6px)] focus:before:rounded-[0.45rem] focus:before:absolute focus:before:inset-0 focus:before:inset-x-[-3px] focus:before:inset-y-[-3px] focus:before:shadow-bottom-gray active:outline-[3px] active:outline-offset-0 active:shadow-inset active:before:content-[""] active:before:w-[calc(100%+6px)] active:before:rounded-[0.45rem] active:before:absolute active:before:inset-0 active:before:inset-x-[-3px] active:before:inset-y-[-3px] active:before:shadow-bottom-white t-secondary-button"
          data-testid="modal-secondary-button"
          type="submit"
        >
          <span />
          Cancel
          <span />
        </button>
      </div>
    </div>
  </div>
</div>
`;
